{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !jupyter nbextension enable --py --sys-prefix widgetsnbextension\n",
    "# !jupyter nbextension enable --py --sys-prefix qgrid\n",
    "\n",
    "\n",
    "# so we can call the g2p service\n",
    "import requests\n",
    "import json\n",
    "\n",
    "# for processing results\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import Counter\n",
    "import datetime\n",
    "\n",
    "\n",
    "# for displays and interactivity\n",
    "import matplotlib.pyplot as plt\n",
    "import qgrid\n",
    "from IPython.display import IFrame    \n",
    "from IPython.display import Javascript\n",
    "from IPython.display import display, Markdown  \n",
    "\n",
    "\n",
    "\n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "\n",
    "def to_df(generator, feature ):\n",
    "    \"\"\" some client side utility code to take the query results and xform it into a data frame \"\"\"   \n",
    "    def to_dict(generator, feature):\n",
    "        for hit in generator:\n",
    "            drugs = [ec.get('term', ec['description']).encode('utf-8') for ec in hit['association']['environmentalContexts']]\n",
    "            description = hit['association']['description'].encode('utf-8')\n",
    "            source = hit['source'].encode('utf-8')\n",
    "            evidence_label = hit['association']['evidence_label']\n",
    "            id = hit['evidence.id']\n",
    "            phenotype = hit['association']['phenotype']['type']['term'].encode('utf-8')\n",
    "            publications = []\n",
    "            matches = []\n",
    "            \n",
    "            \n",
    "            try:\n",
    "                # compare the queried feature vs all the features in the hit.  highlight the matches\n",
    "                f = feature\n",
    "                if f:\n",
    "                    for f2 in hit.get('features', []):\n",
    "                        if not f2:\n",
    "                            continue\n",
    "                        for s in f.get('synonyms', []):\n",
    "                            if s in f2.get('synonyms', []):\n",
    "                                matches.append(s)\n",
    "                        for p in f.get('pathways',[]):\n",
    "                            if p in f2.get('pathways',[]):\n",
    "                                # matches.append(p)\n",
    "                                matches.append('pathway')\n",
    "                        for sp in f.get('swissprots',[]):\n",
    "                            if sp in f2.get('swissprots',[]):\n",
    "                                matches.append(sp)\n",
    "\n",
    "                        if f.get('sequence_ontology', {'name':'X'}) == f2.get('sequence_ontology', {'name':None}):\n",
    "                            matches.append(f['sequence_ontology']['name'])\n",
    "\n",
    "                        if f.get('geneSymbol', None) == f2.get('geneSymbol', None):\n",
    "                            matches.append(f.get('geneSymbol', None))\n",
    "                            \n",
    "                matches = list(set(matches))\n",
    "                \n",
    "                for e in hit['association']['evidence']:\n",
    "                    for p in e['info']['publications']:\n",
    "                        publications.append(p)\n",
    "\n",
    "\n",
    "                yield {'id':id,\n",
    "                       'source': source,\n",
    "                       'evidence_label': evidence_label,\n",
    "                       'drugs': drugs,\n",
    "                       'description': description,\n",
    "                       'phenotype':phenotype,\n",
    "                       'publications':publications,\n",
    "                       'matches': matches\n",
    "                      }\n",
    "            except Exception as e:\n",
    "                print '!!!',e\n",
    "                \n",
    "    \n",
    "    try:        \n",
    "        return pd.DataFrame.from_records( to_dict(generator, feature) , index='id')\n",
    "    except Exception as e:\n",
    "        # might be empty\n",
    "        return pd.DataFrame()\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# chohort / patient's queries\n",
    "# g2p_url = 'http://10.241.39.211:8080/api/v1/features/associations'\n",
    "g2p_url = 'https://dms-dev.compbio.ohsu.edu/api/v1/features/associations'\n",
    "# p115 = [{'description': '', 'start': 27100988L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'ARID1A', 'alt': 'T', 'ref': 'C', 'chromosome': '1', 'name': 'ARID1A '}, {'description': '', 'start': 204494718L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'MDM4', 'alt': 'G', 'ref': 'C', 'chromosome': '1', 'name': 'MDM4 '}, {'description': '', 'start': 30143341L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'ALK', 'alt': 'A', 'ref': 'G', 'chromosome': '2', 'name': 'ALK '}, {'description': '', 'start': 48028096L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'MSH6', 'alt': 'C', 'ref': 'G', 'chromosome': '2', 'name': 'MSH6 '}, {'description': '', 'start': 10084764L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'FANCD2', 'alt': 'G', 'ref': 'C', 'chromosome': '3', 'name': 'FANCD2 '}, {'description': '', 'start': 142185372L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'ATR', 'alt': 'G', 'ref': 'C', 'chromosome': '3', 'name': 'ATR '}, {'description': '', 'start': 178916923L, 'biomarker_type': 'indel', 'referenceName': 'GRCh37', 'geneSymbol': 'PIK3CA', 'alt': 'C', 'ref': 'CCAGTAG', 'chromosome': '3', 'name': 'PIK3CA '}, {'description': '', 'start': 117647566L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'ROS1', 'alt': 'C', 'ref': 'G', 'chromosome': '6', 'name': 'ROS1 '}, {'description': '', 'start': 87285753L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'NTRK2', 'alt': 'G', 'ref': 'C', 'chromosome': '9', 'name': 'NTRK2 '}, {'description': '', 'start': 108106483L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'ATM', 'alt': 'C', 'ref': 'G', 'chromosome': '11', 'name': 'ATM '}, {'description': '', 'start': 45618180L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'FANCM', 'alt': 'G', 'ref': 'C', 'chromosome': '14', 'name': 'FANCM '}, {'description': '', 'start': 45644454L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'FANCM', 'alt': 'C', 'ref': 'G', 'chromosome': '14', 'name': 'FANCM '}, {'description': '', 'start': 2110733L, 'biomarker_type': 'substitution', 'referenceName': 'GRCh37', 'geneSymbol': 'TSC2', 'alt': 'G', 'ref': 'C', 'chromosome': '16', 'name': 'TSC2 '}, {'description': 'GAIN Copy Number Ratio 3.0528', 'start': 39771328L, 'biomarker_type': 'copy_number_variation', 'referenceName': 'GRCh37', 'geneSymbol': 'IDO1', 'alt': '', 'ref': '', 'chromosome': '8', 'name': 'IDO1 '}, {'description': 'GAIN Copy Number Ratio 5.3306', 'start': 39792474L, 'biomarker_type': 'copy_number_variation', 'referenceName': 'GRCh37', 'geneSymbol': 'IDO2', 'alt': '', 'ref': '', 'chromosome': '8', 'name': 'IDO2 '}, {'description': 'LOSS Copy Number Ratio 0.3497', 'start': 21967751L, 'biomarker_type': 'copy_number_variation', 'referenceName': 'GRCh37', 'geneSymbol': 'CDKN2A', 'alt': '', 'ref': '', 'chromosome': '9', 'name': 'CDKN2A '}, {'description': 'GAIN Copy Number Ratio 7.5539', 'start': 38268656L, 'biomarker_type': 'copy_number_variation', 'referenceName': 'GRCh37', 'geneSymbol': 'FGFR1', 'alt': '', 'ref': '', 'chromosome': '8', 'name': 'FGFR1 '}, {'description': 'LOSS Copy Number Ratio 0.5662', 'start': 89623195L, 'biomarker_type': 'copy_number_variation', 'referenceName': 'GRCh37', 'geneSymbol': 'PTEN', 'alt': '', 'ref': '', 'chromosome': '10', 'name': 'PTEN '}, {'description': 'GAIN Copy Number Ratio 2.883', 'start': 69455873L, 'biomarker_type': 'copy_number_variation', 'referenceName': 'GRCh37', 'geneSymbol': 'CCND1', 'alt': '', 'ref': '', 'chromosome': '11', 'name': 'CCND1 '}, {'description': 'GAIN Copy Number Ratio 2.0415', 'start': 128748315L, 'biomarker_type': 'copy_number_variation', 'referenceName': 'GRCh37', 'geneSymbol': 'MYC', 'alt': '', 'ref': '', 'chromosome': '8', 'name': 'MYC '}]  # NOQA\n",
    "# p204 = [{'end': 41266125L, 'name': 'CTNNB1 p.T41I', 'start': 41266125L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CTNNB1', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'catenin (cadherin-associated protein), beta 1, 88kDa'}, {'end': 142188239L, 'name': 'ATR p.I2164M', 'start': 142188239L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'ATR', 'alt': 'C', 'ref': 'T', 'chromosome': '3', 'description': 'ATR serine/threonine kinase'}, {'end': 89685310L, 'name': 'PTEN p.-70fs', 'start': 89685309L, 'biomarker_type': 'Frame_Shift_Ins', 'referenceName': 'GRCh37', 'geneSymbol': 'PTEN', 'alt': 'AATCT', 'ref': '-', 'chromosome': '10', 'description': 'phosphatase and tensin homolog'}, {'end': 41266125L, 'name': 'CTNNB1 p.T41I', 'start': 41266125L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CTNNB1', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'catenin (cadherin-associated protein), beta 1, 88kDa'}, {'end': 1584977L, 'name': 'P2RY8 p.A159T', 'start': 1584977L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'P2RY8', 'alt': 'T', 'ref': 'C', 'chromosome': 'X', 'description': 'purinergic receptor P2Y, G-protein coupled, 8'}, {'end': 170802031L, 'name': 'TNIK p.T1028A', 'start': 170802031L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'TNIK', 'alt': 'C', 'ref': 'T', 'chromosome': '3', 'description': 'TRAF2 and NCK interacting kinase'}, {'end': 118764345L, 'name': 'CXCR5 p.S31F', 'start': 118764345L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CXCR5', 'alt': 'T', 'ref': 'C', 'chromosome': '11', 'description': 'chemokine (C-X-C motif) receptor 5'}, {'end': 136071095L, 'name': 'ZRANB3 p.L310F', 'start': 136071095L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'ZRANB3', 'alt': 'G', 'ref': 'C', 'chromosome': '2', 'description': 'zinc finger, RAN-binding domain containing 3'}, {'end': 228566043L, 'name': 'OBSCN p.V8839A', 'start': 228566043L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'OBSCN', 'alt': 'C', 'ref': 'T', 'chromosome': '1', 'description': 'obscurin, cytoskeletal calmodulin and titin-interacting RhoGEF'}, {'end': 2097401L, 'name': 'STK35 p.L328V', 'start': 2097401L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'STK35', 'alt': 'G', 'ref': 'C', 'chromosome': '20', 'description': 'serine/threonine kinase 35'}, {'end': 58130871L, 'name': 'AGAP2 p.A387T', 'start': 58130871L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'AGAP2', 'alt': 'T', 'ref': 'C', 'chromosome': '12', 'description': 'ArfGAP with GTPase domain, ankyrin repeat and PH domain 2'}, {'end': 33690771L, 'name': 'IP6K3 p.R320H', 'start': 33690771L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'IP6K3', 'alt': 'T', 'ref': 'C', 'chromosome': '6', 'description': 'inositol hexakisphosphate kinase 3'}, {'end': 101910837L, 'name': 'NALCN p.G408A', 'start': 101910837L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'NALCN', 'alt': 'G', 'ref': 'C', 'chromosome': '13', 'description': 'sodium leak channel, non-selective'}, {'end': 57424051L, 'name': 'MYO1A p.R845K', 'start': 57424051L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'MYO1A', 'alt': 'T', 'ref': 'C', 'chromosome': '12', 'description': 'myosin IA'}, {'end': 38211533L, 'name': 'TRPC4 p.R814K', 'start': 38211533L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'TRPC4', 'alt': 'T', 'ref': 'C', 'chromosome': '13', 'description': 'transient receptor potential cation channel, subfamily C, member 4'}, {'end': 32257834L, 'name': 'SPOCD1 p.R982C', 'start': 32257834L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'SPOCD1', 'alt': 'A', 'ref': 'G', 'chromosome': '1', 'description': 'SPOC domain containing 1'}, {'end': 28097282L, 'name': 'ZSCAN16 p.E201K', 'start': 28097282L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'ZSCAN16', 'alt': 'A', 'ref': 'G', 'chromosome': '6', 'description': 'zinc finger and SCAN domain containing 16'}, {'end': 84628811L, 'name': 'SEMA3D p.R760Q', 'start': 84628811L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'SEMA3D', 'alt': 'T', 'ref': 'C', 'chromosome': '7', 'description': 'sema domain, immunoglobulin domain (Ig), short basic domain, secreted, (semaphorin) 3D'}, {'end': 72400932L, 'name': 'NEGR1 p.A80V', 'start': 72400932L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'NEGR1', 'alt': 'A', 'ref': 'G', 'chromosome': '1', 'description': 'neuronal growth regulator 1'}, {'end': 35778772L, 'name': 'ARPP21 p.P521L', 'start': 35778772L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'ARPP21', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'cAMP-regulated phosphoprotein, 21kDa'}, {'end': 21016738L, 'name': 'KIF17 p.R442S', 'start': 21016738L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'KIF17', 'alt': 'T', 'ref': 'G', 'chromosome': '1', 'description': 'kinesin family member 17'}, {'end': 51929267L, 'name': 'IQCF1 p.R86Q', 'start': 51929267L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'IQCF1', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'IQ motif containing F1'}, {'end': 95188799L, 'name': 'CDH17 p.E132K', 'start': 95188799L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CDH17', 'alt': 'T', 'ref': 'C', 'chromosome': '8', 'description': 'cadherin 17, LI cadherin (liver-intestine)'}, {'end': 96117712L, 'name': 'CCDC82 p.L67P', 'start': 96117712L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CCDC82', 'alt': 'G', 'ref': 'A', 'chromosome': '11', 'description': 'coiled-coil domain containing 82'}, {'end': 105967614L, 'name': 'AASDHPPT p.R304*', 'start': 105967614L, 'biomarker_type': 'Nonsense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'AASDHPPT', 'alt': 'T', 'ref': 'C', 'chromosome': '11', 'description': 'aminoadipate-semialdehyde dehydrogenase-phosphopantetheinyl transferase'}, {'end': 31538698L, 'name': 'CLDN17 p.A80S', 'start': 31538698L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CLDN17', 'alt': 'A', 'ref': 'C', 'chromosome': '21', 'description': 'claudin 17'}, {'end': 38483424L, 'name': 'UTP11L p.M70I', 'start': 38483424L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'UTP11L', 'alt': 'A', 'ref': 'G', 'chromosome': '1', 'description': 'UTP11-like, U3 small nucleolar ribonucleoprotein (yeast)'}, {'end': 155526050L, 'name': 'FGG p.Q433P', 'start': 155526050L, 'biomarker_type': 'Splice_Site', 'referenceName': 'GRCh37', 'geneSymbol': 'FGG', 'alt': 'G', 'ref': 'T', 'chromosome': '4', 'description': 'fibrinogen gamma chain'}, {'end': 148617051L, 'name': 'ABLIM3 p.A310V', 'start': 148617051L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'ABLIM3', 'alt': 'T', 'ref': 'C', 'chromosome': '5', 'description': 'actin binding LIM protein family, member 3'}, {'end': 31895941L, 'name': 'CFB p.P86S', 'start': 31895941L, 'biomarker_type': 'Splice_Site', 'referenceName': 'GRCh37', 'geneSymbol': 'CFB', 'alt': 'T', 'ref': 'C', 'chromosome': '6', 'description': 'complement factor B'}, {'end': 18051501L, 'name': 'MYO15A p.A2223V', 'start': 18051501L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'MYO15A', 'alt': 'T', 'ref': 'C', 'chromosome': '17', 'description': 'myosin XVA'}, {'end': 60969271L, 'name': 'CABLES2 p.S219C', 'start': 60969271L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CABLES2', 'alt': 'C', 'ref': 'G', 'chromosome': '20', 'description': 'Cdk5 and Abl enzyme substrate 2'}, {'end': 186096984L, 'name': 'KIAA1430 p.L426F', 'start': 186096984L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'KIAA1430', 'alt': 'A', 'ref': 'G', 'chromosome': '4', 'description': ''}, {'end': 110056501L, 'name': 'FIG4 p.G216R', 'start': 110056501L, 'biomarker_type': 'Splice_Site', 'referenceName': 'GRCh37', 'geneSymbol': 'FIG4', 'alt': 'A', 'ref': 'G', 'chromosome': '6', 'description': 'FIG4 phosphoinositide 5-phosphatase'}, {'end': 67690336L, 'name': 'RLTPR p.C1275Y', 'start': 67690336L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'RLTPR', 'alt': 'A', 'ref': 'G', 'chromosome': '16', 'description': 'RGD motif, leucine rich repeats, tropomodulin domain and proline-rich containing'}, {'end': 67702418L, 'name': 'C16orf86 p.D290G', 'start': 67702418L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'C16orf86', 'alt': 'G', 'ref': 'A', 'chromosome': '16', 'description': 'chromosome 16 open reading frame 86'}]\n",
    "\n",
    "# p115 = [{'description': '', 'start': 27100988L, 'biomarker_type': 'substitution', 'geneSymbol': 'ARID1A', 'alt': 'T', 'ref': 'C', 'chromosome': '1', 'name': 'ARID1A '}, {'description': '', 'start': 204494718L, 'biomarker_type': 'substitution', 'geneSymbol': 'MDM4', 'alt': 'G', 'ref': 'C', 'chromosome': '1', 'name': 'MDM4 '}, {'description': '', 'start': 30143341L, 'biomarker_type': 'substitution', 'geneSymbol': 'ALK', 'alt': 'A', 'ref': 'G', 'chromosome': '2', 'name': 'ALK '}, {'description': '', 'start': 48028096L, 'biomarker_type': 'substitution', 'geneSymbol': 'MSH6', 'alt': 'C', 'ref': 'G', 'chromosome': '2', 'name': 'MSH6 '}, {'description': '', 'start': 10084764L, 'biomarker_type': 'substitution', 'geneSymbol': 'FANCD2', 'alt': 'G', 'ref': 'C', 'chromosome': '3', 'name': 'FANCD2 '}, {'description': '', 'start': 142185372L, 'biomarker_type': 'substitution', 'geneSymbol': 'ATR', 'alt': 'G', 'ref': 'C', 'chromosome': '3', 'name': 'ATR '}, {'description': '', 'start': 178916923L, 'biomarker_type': 'indel', 'geneSymbol': 'PIK3CA', 'alt': 'C', 'ref': 'CCAGTAG', 'chromosome': '3', 'name': 'PIK3CA '}, {'description': '', 'start': 117647566L, 'biomarker_type': 'substitution', 'geneSymbol': 'ROS1', 'alt': 'C', 'ref': 'G', 'chromosome': '6', 'name': 'ROS1 '}, {'description': '', 'start': 87285753L, 'biomarker_type': 'substitution', 'geneSymbol': 'NTRK2', 'alt': 'G', 'ref': 'C', 'chromosome': '9', 'name': 'NTRK2 '}, {'description': '', 'start': 108106483L, 'biomarker_type': 'substitution', 'geneSymbol': 'ATM', 'alt': 'C', 'ref': 'G', 'chromosome': '11', 'name': 'ATM '}, {'description': '', 'start': 45618180L, 'biomarker_type': 'substitution', 'geneSymbol': 'FANCM', 'alt': 'G', 'ref': 'C', 'chromosome': '14', 'name': 'FANCM '}, {'description': '', 'start': 45644454L, 'biomarker_type': 'substitution', 'geneSymbol': 'FANCM', 'alt': 'C', 'ref': 'G', 'chromosome': '14', 'name': 'FANCM '}, {'description': '', 'start': 2110733L, 'biomarker_type': 'substitution', 'geneSymbol': 'TSC2', 'alt': 'G', 'ref': 'C', 'chromosome': '16', 'name': 'TSC2 '}, {'description': 'GAIN Copy Number Ratio 3.0528', 'start': 39771328L, 'biomarker_type': 'copy_number_variation', 'geneSymbol': 'IDO1', 'alt': '', 'ref': '', 'chromosome': '8', 'name': 'IDO1 '}, {'description': 'GAIN Copy Number Ratio 5.3306', 'start': 39792474L, 'biomarker_type': 'copy_number_variation', 'geneSymbol': 'IDO2', 'alt': '', 'ref': '', 'chromosome': '8', 'name': 'IDO2 '}, {'description': 'LOSS Copy Number Ratio 0.3497', 'start': 21967751L, 'biomarker_type': 'copy_number_variation', 'geneSymbol': 'CDKN2A', 'alt': '', 'ref': '', 'chromosome': '9', 'name': 'CDKN2A '}, {'description': 'GAIN Copy Number Ratio 7.5539', 'start': 38268656L, 'biomarker_type': 'copy_number_variation', 'geneSymbol': 'FGFR1', 'alt': '', 'ref': '', 'chromosome': '8', 'name': 'FGFR1 '}, {'description': 'LOSS Copy Number Ratio 0.5662', 'start': 89623195L, 'biomarker_type': 'copy_number_variation', 'geneSymbol': 'PTEN', 'alt': '', 'ref': '', 'chromosome': '10', 'name': 'PTEN '}, {'description': 'GAIN Copy Number Ratio 2.883', 'start': 69455873L, 'biomarker_type': 'copy_number_variation', 'geneSymbol': 'CCND1', 'alt': '', 'ref': '', 'chromosome': '11', 'name': 'CCND1 '}, {'description': 'GAIN Copy Number Ratio 2.0415', 'start': 128748315L, 'biomarker_type': 'copy_number_variation', 'geneSymbol': 'MYC', 'alt': '', 'ref': '', 'chromosome': '8', 'name': 'MYC '}]  # NOQA\n",
    "# p204 = [{'end': 41266125L, 'name': 'CTNNB1 p.T41I', 'start': 41266125L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'CTNNB1', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'catenin (cadherin-associated protein), beta 1, 88kDa'}, {'end': 142188239L, 'name': 'ATR p.I2164M', 'start': 142188239L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'ATR', 'alt': 'C', 'ref': 'T', 'chromosome': '3', 'description': 'ATR serine/threonine kinase'}, {'end': 89685310L, 'name': 'PTEN p.-70fs', 'start': 89685309L, 'biomarker_type': 'Frame_Shift_Ins', 'geneSymbol': 'PTEN', 'alt': 'AATCT', 'ref': '-', 'chromosome': '10', 'description': 'phosphatase and tensin homolog'}, {'end': 41266125L, 'name': 'CTNNB1 p.T41I', 'start': 41266125L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'CTNNB1', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'catenin (cadherin-associated protein), beta 1, 88kDa'}, {'end': 1584977L, 'name': 'P2RY8 p.A159T', 'start': 1584977L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'P2RY8', 'alt': 'T', 'ref': 'C', 'chromosome': 'X', 'description': 'purinergic receptor P2Y, G-protein coupled, 8'}, {'end': 170802031L, 'name': 'TNIK p.T1028A', 'start': 170802031L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'TNIK', 'alt': 'C', 'ref': 'T', 'chromosome': '3', 'description': 'TRAF2 and NCK interacting kinase'}, {'end': 118764345L, 'name': 'CXCR5 p.S31F', 'start': 118764345L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'CXCR5', 'alt': 'T', 'ref': 'C', 'chromosome': '11', 'description': 'chemokine (C-X-C motif) receptor 5'}, {'end': 136071095L, 'name': 'ZRANB3 p.L310F', 'start': 136071095L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'ZRANB3', 'alt': 'G', 'ref': 'C', 'chromosome': '2', 'description': 'zinc finger, RAN-binding domain containing 3'}, {'end': 228566043L, 'name': 'OBSCN p.V8839A', 'start': 228566043L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'OBSCN', 'alt': 'C', 'ref': 'T', 'chromosome': '1', 'description': 'obscurin, cytoskeletal calmodulin and titin-interacting RhoGEF'}, {'end': 2097401L, 'name': 'STK35 p.L328V', 'start': 2097401L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'STK35', 'alt': 'G', 'ref': 'C', 'chromosome': '20', 'description': 'serine/threonine kinase 35'}, {'end': 58130871L, 'name': 'AGAP2 p.A387T', 'start': 58130871L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'AGAP2', 'alt': 'T', 'ref': 'C', 'chromosome': '12', 'description': 'ArfGAP with GTPase domain, ankyrin repeat and PH domain 2'}, {'end': 33690771L, 'name': 'IP6K3 p.R320H', 'start': 33690771L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'IP6K3', 'alt': 'T', 'ref': 'C', 'chromosome': '6', 'description': 'inositol hexakisphosphate kinase 3'}, {'end': 101910837L, 'name': 'NALCN p.G408A', 'start': 101910837L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'NALCN', 'alt': 'G', 'ref': 'C', 'chromosome': '13', 'description': 'sodium leak channel, non-selective'}, {'end': 57424051L, 'name': 'MYO1A p.R845K', 'start': 57424051L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'MYO1A', 'alt': 'T', 'ref': 'C', 'chromosome': '12', 'description': 'myosin IA'}, {'end': 38211533L, 'name': 'TRPC4 p.R814K', 'start': 38211533L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'TRPC4', 'alt': 'T', 'ref': 'C', 'chromosome': '13', 'description': 'transient receptor potential cation channel, subfamily C, member 4'}, {'end': 32257834L, 'name': 'SPOCD1 p.R982C', 'start': 32257834L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'SPOCD1', 'alt': 'A', 'ref': 'G', 'chromosome': '1', 'description': 'SPOC domain containing 1'}, {'end': 28097282L, 'name': 'ZSCAN16 p.E201K', 'start': 28097282L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'ZSCAN16', 'alt': 'A', 'ref': 'G', 'chromosome': '6', 'description': 'zinc finger and SCAN domain containing 16'}, {'end': 84628811L, 'name': 'SEMA3D p.R760Q', 'start': 84628811L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'SEMA3D', 'alt': 'T', 'ref': 'C', 'chromosome': '7', 'description': 'sema domain, immunoglobulin domain (Ig), short basic domain, secreted, (semaphorin) 3D'}, {'end': 72400932L, 'name': 'NEGR1 p.A80V', 'start': 72400932L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'NEGR1', 'alt': 'A', 'ref': 'G', 'chromosome': '1', 'description': 'neuronal growth regulator 1'}, {'end': 35778772L, 'name': 'ARPP21 p.P521L', 'start': 35778772L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'ARPP21', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'cAMP-regulated phosphoprotein, 21kDa'}, {'end': 21016738L, 'name': 'KIF17 p.R442S', 'start': 21016738L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'KIF17', 'alt': 'T', 'ref': 'G', 'chromosome': '1', 'description': 'kinesin family member 17'}, {'end': 51929267L, 'name': 'IQCF1 p.R86Q', 'start': 51929267L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'IQCF1', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'IQ motif containing F1'}, {'end': 95188799L, 'name': 'CDH17 p.E132K', 'start': 95188799L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'CDH17', 'alt': 'T', 'ref': 'C', 'chromosome': '8', 'description': 'cadherin 17, LI cadherin (liver-intestine)'}, {'end': 96117712L, 'name': 'CCDC82 p.L67P', 'start': 96117712L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'CCDC82', 'alt': 'G', 'ref': 'A', 'chromosome': '11', 'description': 'coiled-coil domain containing 82'}, {'end': 105967614L, 'name': 'AASDHPPT p.R304*', 'start': 105967614L, 'biomarker_type': 'Nonsense_Mutation', 'geneSymbol': 'AASDHPPT', 'alt': 'T', 'ref': 'C', 'chromosome': '11', 'description': 'aminoadipate-semialdehyde dehydrogenase-phosphopantetheinyl transferase'}, {'end': 31538698L, 'name': 'CLDN17 p.A80S', 'start': 31538698L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'CLDN17', 'alt': 'A', 'ref': 'C', 'chromosome': '21', 'description': 'claudin 17'}, {'end': 38483424L, 'name': 'UTP11L p.M70I', 'start': 38483424L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'UTP11L', 'alt': 'A', 'ref': 'G', 'chromosome': '1', 'description': 'UTP11-like, U3 small nucleolar ribonucleoprotein (yeast)'}, {'end': 155526050L, 'name': 'FGG p.Q433P', 'start': 155526050L, 'biomarker_type': 'Splice_Site', 'geneSymbol': 'FGG', 'alt': 'G', 'ref': 'T', 'chromosome': '4', 'description': 'fibrinogen gamma chain'}, {'end': 148617051L, 'name': 'ABLIM3 p.A310V', 'start': 148617051L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'ABLIM3', 'alt': 'T', 'ref': 'C', 'chromosome': '5', 'description': 'actin binding LIM protein family, member 3'}, {'end': 31895941L, 'name': 'CFB p.P86S', 'start': 31895941L, 'biomarker_type': 'Splice_Site', 'geneSymbol': 'CFB', 'alt': 'T', 'ref': 'C', 'chromosome': '6', 'description': 'complement factor B'}, {'end': 18051501L, 'name': 'MYO15A p.A2223V', 'start': 18051501L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'MYO15A', 'alt': 'T', 'ref': 'C', 'chromosome': '17', 'description': 'myosin XVA'}, {'end': 60969271L, 'name': 'CABLES2 p.S219C', 'start': 60969271L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'CABLES2', 'alt': 'C', 'ref': 'G', 'chromosome': '20', 'description': 'Cdk5 and Abl enzyme substrate 2'}, {'end': 186096984L, 'name': 'KIAA1430 p.L426F', 'start': 186096984L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'KIAA1430', 'alt': 'A', 'ref': 'G', 'chromosome': '4', 'description': ''}, {'end': 110056501L, 'name': 'FIG4 p.G216R', 'start': 110056501L, 'biomarker_type': 'Splice_Site', 'geneSymbol': 'FIG4', 'alt': 'A', 'ref': 'G', 'chromosome': '6', 'description': 'FIG4 phosphoinositide 5-phosphatase'}, {'end': 67690336L, 'name': 'RLTPR p.C1275Y', 'start': 67690336L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'RLTPR', 'alt': 'A', 'ref': 'G', 'chromosome': '16', 'description': 'RGD motif, leucine rich repeats, tropomodulin domain and proline-rich containing'}, {'end': 67702418L, 'name': 'C16orf86 p.D290G', 'start': 67702418L, 'biomarker_type': 'Missense_Mutation', 'geneSymbol': 'C16orf86', 'alt': 'G', 'ref': 'A', 'chromosome': '16', 'description': 'chromosome 16 open reading frame 86'}]\n",
    "\n",
    "p204 = [{'end': 41266125L, 'name': 'CTNNB1 p.T41I', 'start': 41266125L, 'biomarker_type': 'Missense_Mutation', 'referenceName': '3', 'geneSymbol': 'CTNNB1', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'catenin (cadherin-associated protein), beta 1, 88kDa'}, {'end': 142188239L, 'name': 'ATR p.I2164M', 'start': 142188239L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'ATR', 'alt': 'C', 'ref': 'T', 'chromosome': '3', 'description': 'ATR serine/threonine kinase'}, {'end': 89685310L, 'name': 'PTEN p.-70fs', 'start': 89685309L, 'biomarker_type': 'Frame_Shift_Ins', 'referenceName': 'GRCh37', 'geneSymbol': 'PTEN', 'alt': 'AATCT', 'ref': '-', 'chromosome': '10', 'description': 'phosphatase and tensin homolog'}, {'end': 41266125L, 'name': 'CTNNB1 p.T41I', 'start': 41266125L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CTNNB1', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'catenin (cadherin-associated protein), beta 1, 88kDa'}, {'end': 1584977L, 'name': 'P2RY8 p.A159T', 'start': 1584977L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'P2RY8', 'alt': 'T', 'ref': 'C', 'chromosome': 'X', 'description': 'purinergic receptor P2Y, G-protein coupled, 8'}, {'end': 170802031L, 'name': 'TNIK p.T1028A', 'start': 170802031L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'TNIK', 'alt': 'C', 'ref': 'T', 'chromosome': '3', 'description': 'TRAF2 and NCK interacting kinase'}, {'end': 118764345L, 'name': 'CXCR5 p.S31F', 'start': 118764345L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CXCR5', 'alt': 'T', 'ref': 'C', 'chromosome': '11', 'description': 'chemokine (C-X-C motif) receptor 5'}, {'end': 136071095L, 'name': 'ZRANB3 p.L310F', 'start': 136071095L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'ZRANB3', 'alt': 'G', 'ref': 'C', 'chromosome': '2', 'description': 'zinc finger, RAN-binding domain containing 3'}, {'end': 228566043L, 'name': 'OBSCN p.V8839A', 'start': 228566043L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'OBSCN', 'alt': 'C', 'ref': 'T', 'chromosome': '1', 'description': 'obscurin, cytoskeletal calmodulin and titin-interacting RhoGEF'}, {'end': 2097401L, 'name': 'STK35 p.L328V', 'start': 2097401L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'STK35', 'alt': 'G', 'ref': 'C', 'chromosome': '20', 'description': 'serine/threonine kinase 35'}, {'end': 58130871L, 'name': 'AGAP2 p.A387T', 'start': 58130871L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'AGAP2', 'alt': 'T', 'ref': 'C', 'chromosome': '12', 'description': 'ArfGAP with GTPase domain, ankyrin repeat and PH domain 2'}, {'end': 33690771L, 'name': 'IP6K3 p.R320H', 'start': 33690771L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'IP6K3', 'alt': 'T', 'ref': 'C', 'chromosome': '6', 'description': 'inositol hexakisphosphate kinase 3'}, {'end': 101910837L, 'name': 'NALCN p.G408A', 'start': 101910837L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'NALCN', 'alt': 'G', 'ref': 'C', 'chromosome': '13', 'description': 'sodium leak channel, non-selective'}, {'end': 57424051L, 'name': 'MYO1A p.R845K', 'start': 57424051L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'MYO1A', 'alt': 'T', 'ref': 'C', 'chromosome': '12', 'description': 'myosin IA'}, {'end': 38211533L, 'name': 'TRPC4 p.R814K', 'start': 38211533L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'TRPC4', 'alt': 'T', 'ref': 'C', 'chromosome': '13', 'description': 'transient receptor potential cation channel, subfamily C, member 4'}, {'end': 32257834L, 'name': 'SPOCD1 p.R982C', 'start': 32257834L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'SPOCD1', 'alt': 'A', 'ref': 'G', 'chromosome': '1', 'description': 'SPOC domain containing 1'}, {'end': 28097282L, 'name': 'ZSCAN16 p.E201K', 'start': 28097282L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'ZSCAN16', 'alt': 'A', 'ref': 'G', 'chromosome': '6', 'description': 'zinc finger and SCAN domain containing 16'}, {'end': 84628811L, 'name': 'SEMA3D p.R760Q', 'start': 84628811L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'SEMA3D', 'alt': 'T', 'ref': 'C', 'chromosome': '7', 'description': 'sema domain, immunoglobulin domain (Ig), short basic domain, secreted, (semaphorin) 3D'}, {'end': 72400932L, 'name': 'NEGR1 p.A80V', 'start': 72400932L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'NEGR1', 'alt': 'A', 'ref': 'G', 'chromosome': '1', 'description': 'neuronal growth regulator 1'}, {'end': 35778772L, 'name': 'ARPP21 p.P521L', 'start': 35778772L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'ARPP21', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'cAMP-regulated phosphoprotein, 21kDa'}, {'end': 21016738L, 'name': 'KIF17 p.R442S', 'start': 21016738L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'KIF17', 'alt': 'T', 'ref': 'G', 'chromosome': '1', 'description': 'kinesin family member 17'}, {'end': 51929267L, 'name': 'IQCF1 p.R86Q', 'start': 51929267L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'IQCF1', 'alt': 'T', 'ref': 'C', 'chromosome': '3', 'description': 'IQ motif containing F1'}, {'end': 95188799L, 'name': 'CDH17 p.E132K', 'start': 95188799L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CDH17', 'alt': 'T', 'ref': 'C', 'chromosome': '8', 'description': 'cadherin 17, LI cadherin (liver-intestine)'}, {'end': 96117712L, 'name': 'CCDC82 p.L67P', 'start': 96117712L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CCDC82', 'alt': 'G', 'ref': 'A', 'chromosome': '11', 'description': 'coiled-coil domain containing 82'}, {'end': 105967614L, 'name': 'AASDHPPT p.R304*', 'start': 105967614L, 'biomarker_type': 'Nonsense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'AASDHPPT', 'alt': 'T', 'ref': 'C', 'chromosome': '11', 'description': 'aminoadipate-semialdehyde dehydrogenase-phosphopantetheinyl transferase'}, {'end': 31538698L, 'name': 'CLDN17 p.A80S', 'start': 31538698L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CLDN17', 'alt': 'A', 'ref': 'C', 'chromosome': '21', 'description': 'claudin 17'}, {'end': 38483424L, 'name': 'UTP11L p.M70I', 'start': 38483424L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'UTP11L', 'alt': 'A', 'ref': 'G', 'chromosome': '1', 'description': 'UTP11-like, U3 small nucleolar ribonucleoprotein (yeast)'}, {'end': 155526050L, 'name': 'FGG p.Q433P', 'start': 155526050L, 'biomarker_type': 'Splice_Site', 'referenceName': 'GRCh37', 'geneSymbol': 'FGG', 'alt': 'G', 'ref': 'T', 'chromosome': '4', 'description': 'fibrinogen gamma chain'}, {'end': 148617051L, 'name': 'ABLIM3 p.A310V', 'start': 148617051L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'ABLIM3', 'alt': 'T', 'ref': 'C', 'chromosome': '5', 'description': 'actin binding LIM protein family, member 3'}, {'end': 31895941L, 'name': 'CFB p.P86S', 'start': 31895941L, 'biomarker_type': 'Splice_Site', 'referenceName': 'GRCh37', 'geneSymbol': 'CFB', 'alt': 'T', 'ref': 'C', 'chromosome': '6', 'description': 'complement factor B'}, {'end': 18051501L, 'name': 'MYO15A p.A2223V', 'start': 18051501L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'MYO15A', 'alt': 'T', 'ref': 'C', 'chromosome': '17', 'description': 'myosin XVA'}, {'end': 60969271L, 'name': 'CABLES2 p.S219C', 'start': 60969271L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'CABLES2', 'alt': 'C', 'ref': 'G', 'chromosome': '20', 'description': 'Cdk5 and Abl enzyme substrate 2'}, {'end': 186096984L, 'name': 'KIAA1430 p.L426F', 'start': 186096984L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'KIAA1430', 'alt': 'A', 'ref': 'G', 'chromosome': '4', 'description': ''}, {'end': 110056501L, 'name': 'FIG4 p.G216R', 'start': 110056501L, 'biomarker_type': 'Splice_Site', 'referenceName': 'GRCh37', 'geneSymbol': 'FIG4', 'alt': 'A', 'ref': 'G', 'chromosome': '6', 'description': 'FIG4 phosphoinositide 5-phosphatase'}, {'end': 67690336L, 'name': 'RLTPR p.C1275Y', 'start': 67690336L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'RLTPR', 'alt': 'A', 'ref': 'G', 'chromosome': '16', 'description': 'RGD motif, leucine rich repeats, tropomodulin domain and proline-rich containing'}, {'end': 67702418L, 'name': 'C16orf86 p.D290G', 'start': 67702418L, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'C16orf86', 'alt': 'G', 'ref': 'A', 'chromosome': '16', 'description': 'chromosome 16 open reading frame 86'}]\n",
    "\n",
    "test = [{\"start\": 41266125, \"end\": 41266125, \"name\": \"CTNNB1 p.T41I\", \"referenceName\": \"3\", \"geneSymbol\": \"CTNNB1\", \"alt\": \"T\", \"description\": \"catenin (cadherin-associated protein), beta 1, 88kDa\", \"ref\": \"C\", \"chromosome\": \"3\", \"biomarker_type\": \"Missense_Mutation\"}]\n",
    "test = [{'end': 228566043, 'name': 'OBSCN p.V8839A', 'start': 228566043, 'biomarker_type': 'Missense_Mutation', 'referenceName': 'GRCh37', 'geneSymbol': 'OBSCN', 'alt': 'C', 'ref': 'T', 'chromosome': '1', 'description': 'obscurin, cytoskeletal calmodulin and titin-interacting RhoGEF'}]\n",
    "\n",
    "dataset_features = test\n",
    "\n",
    "dataset_name = 'test'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[{\"start\": 228566043, \"end\": 228566043, \"name\": \"OBSCN p.V8839A\", \"referenceName\": \"GRCh37\", \"geneSymbol\": \"OBSCN\", \"alt\": \"C\", \"description\": \"obscurin, cytoskeletal calmodulin and titin-interacting RhoGEF\", \"ref\": \"T\", \"chromosome\": \"1\", \"biomarker_type\": \"Missense_Mutation\"}]'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json.dumps(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total of 8 queries, 1041 total hits, 1041 unique hits returned in 0:00:12.859547\n"
     ]
    }
   ],
   "source": [
    "# run the queries and create data frames\n",
    "\n",
    "queries = []\n",
    "data_frames = {}\n",
    "t = datetime.datetime.now()     \n",
    "\n",
    "rsp = requests.post(g2p_url, json=dataset_features, timeout=240)\n",
    "\n",
    "location_queries = rsp.json()    \n",
    "\n",
    "evidence_ids = []\n",
    "\n",
    "for location_query in location_queries:\n",
    "    allele = location_query['allele']\n",
    "    name = location_query['name']\n",
    "    biomarker_type = location_query['biomarker_type']\n",
    "    if allele not in data_frames:\n",
    "        data_frames[allele] = {}\n",
    "    data_frames[allele][name] = to_df(location_query['hits'],location_query['feature'])\n",
    "    # now that we have the associations in a dataframe, lets recast the hits to a hit count\n",
    "    evidence_ids.extend([hit['evidence.id'] for hit in location_query['hits']])\n",
    "    location_query['hits'] = len(location_query['hits'])\n",
    "    queries.append(location_query)\n",
    "\n",
    "print 'total of {} queries, {} total hits, {} unique hits returned in {}'.format(\n",
    "    len(queries),\n",
    "    len(evidence_ids),\n",
    "    len(list(set(evidence_ids))),\n",
    "    datetime.datetime.now() - t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9d5078c5714742e69df555c495acf24c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>QgridWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "QgridWidget(grid_options={'defaultColumnWidth': 150, 'highlightSelectedRow': True, 'enableTextSelectionOnCells': True, 'editable': False, 'explicitInitialization': True, 'minVisibleRows': 8, 'enableColumnReorder': False, 'maxVisibleRows': 15, 'sortable': True, 'fullWidthRows': True, 'rowHeight': 28, 'filterable': True, 'autoEdit': False, 'syncColumnCellResize': True, 'forceFitColumns': True, 'highlightSelectedCell': False}, precision=5)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# summarize the queries by allele\n",
    "\n",
    "queries_summary = {}\n",
    "\n",
    "for q in queries:\n",
    "    if q['allele'] not in queries_summary:\n",
    "        queries_summary[q['allele']] = { 'allele': q['allele'], 'biomarker_type': q['biomarker_type'] }\n",
    "    queries_summary[q['allele']]['hits-{}'.format(q['name'])] = q['hits']    \n",
    "\n",
    "df = pd.DataFrame.from_records([queries_summary[k] for k in queries_summary.keys()]).fillna(0)\n",
    "queries_summary_df_widget = qgrid.show_grid(df, grid_options={'editable':False})\n",
    "display(queries_summary_df_widget)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<p>query summary:</p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "879e13fe51514f63bb6bbf41c8d63acd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>QgridWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "QgridWidget(grid_options={'defaultColumnWidth': 150, 'highlightSelectedRow': True, 'enableTextSelectionOnCells': True, 'editable': False, 'explicitInitialization': True, 'minVisibleRows': 8, 'enableColumnReorder': False, 'maxVisibleRows': 15, 'sortable': True, 'fullWidthRows': True, 'rowHeight': 28, 'filterable': True, 'autoEdit': False, 'syncColumnCellResize': True, 'forceFitColumns': True, 'highlightSelectedCell': False}, precision=5)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<p>top3:</p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4738644e405e4dc8b50c754a4dcbb6ee",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>QgridWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "QgridWidget(grid_options={'highlightSelectedRow': True, 'fullWidthRows': True, 'rowHeight': 28, 'enableColumnReorder': False, 'filterable': True, 'sortable': True, 'enableTextSelectionOnCells': True, 'editable': True, 'explicitInitialization': True, 'minVisibleRows': 8, 'autoEdit': False, 'syncColumnCellResize': True, 'maxVisibleRows': 15, 'defaultColumnWidth': 150, 'forceFitColumns': True, 'highlightSelectedCell': False}, precision=5)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<p>OBSCN NC_000001.10:g.228566043T>C protein_domains</p><p>-source:*trials +association.environmentalContexts.id:('CID23725625', 'CID56842121', u'CHEMBL3137343', 'CID5330286', 'CID444795', 'CID10184653', 'CID5311', 'CID6442177', 'CID11707110', 'CID25102847', 'CID9823820', 'CID24826799', u'CHEMBL1789844', u'CHEMBL2108738', u'CHEMBL2007641', u'CHEMBL1351', 'CID15951529', 'CID132971', 'CID42611257', 'CID9854073', 'CID6918837', 'CID5291', 'CID3062316', 'CID5329102', 'CID216239', 'CID25126798', 'CID387447', 'CID11625818', 'CID49846579', 'CID5284616', u'CHEMBL1201583', 'CID176870', 'CID2662') +features.protein_domains.name:('1' OR 'PTHR19897' OR 'PF00069' OR 'SSF56112' OR 'PS50011' OR 'SM00220')</p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e48a375c4b0d4042867f4a9c728b21f8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>QgridWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "QgridWidget(grid_options={'highlightSelectedRow': True, 'fullWidthRows': True, 'rowHeight': 28, 'enableColumnReorder': False, 'filterable': True, 'sortable': True, 'enableTextSelectionOnCells': True, 'editable': True, 'explicitInitialization': True, 'minVisibleRows': 8, 'autoEdit': False, 'syncColumnCellResize': True, 'maxVisibleRows': 15, 'defaultColumnWidth': 150, 'forceFitColumns': True, 'highlightSelectedCell': False}, precision=5)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create an 'interactive' master detail\n",
    "\n",
    "from IPython.display import clear_output\n",
    "from IPython.display import IFrame  \n",
    "from IPython.core.display import HTML\n",
    "\n",
    "queries_df = pd.DataFrame.from_records(queries)\n",
    "queries_df_widget = qgrid.show_grid(queries_df, grid_options={'editable':False})\n",
    "qgrid_widget = None\n",
    "top3_widget = None\n",
    "allele = None \n",
    "name = None\n",
    "\n",
    "def top(df, fields = ['publications', 'evidence_label', 'phenotype', 'drugs' ], size=3):\n",
    "    counters = []\n",
    "    for f in fields:\n",
    "        counter = Counter()\n",
    "        for value in df[f]:\n",
    "            if isinstance(value, (list,tuple,)):\n",
    "                for v in value:\n",
    "                    counter[v] += 1\n",
    "            else:\n",
    "                counter[value] += 1\n",
    "        for c in counter.most_common(size):\n",
    "            counters.append({'field': f, 'value': c[0], 'count': c[1]})\n",
    "    if len(counters) == 0:\n",
    "        counters.append({'field': fields, 'value': None, 'count': None})\n",
    "    return counters    \n",
    "    \n",
    "\n",
    "def on_queries_df_row_selected(change):\n",
    "    global allele\n",
    "    global name\n",
    "    global qgrid_widget\n",
    "    global top3_widget    \n",
    "    \n",
    "    selected_row = queries_df_widget.get_selected_df().to_dict('records')\n",
    "    allele = selected_row[0]['allele']\n",
    "    name = selected_row[0]['name']\n",
    "    query_string = selected_row[0]['query_string']\n",
    "        \n",
    "    top3_widget = None\n",
    "    \n",
    "    try:    \n",
    "        df = data_frames[allele][name]\n",
    "        qgrid_widget = qgrid.show_grid(df)\n",
    "        top3_widget = qgrid.show_grid(pd.DataFrame(top(df)))\n",
    "        \n",
    "    except Exception as e:\n",
    "        # df could be empty\n",
    "        pass  \n",
    "    \n",
    "    # show publication\n",
    "    def on_row_selected(change):\n",
    "        global qgrid_widget\n",
    "        try:\n",
    "            selected_row = qgrid_widget.get_selected_df().to_dict('records')\n",
    "            publications = selected_row[0]['publications']\n",
    "            matches = selected_row[0]['matches']            \n",
    "            clear_output(wait=True)\n",
    "            display(queries_df_widget,\n",
    "                    HTML('<p>{} {}</p><p>{}</p>'.format(allele, name, query_string)),\n",
    "                    qgrid_widget,\n",
    "                    HTML('<p>matches:{}</p><p>{}</p>'.format(matches, publications)),\n",
    "                    IFrame(publications[0], width=900, height=900))    \n",
    "        except Exception as e:          \n",
    "            print e\n",
    "\n",
    "    qgrid_widget.observe(on_row_selected, names=['_selected_rows'])        \n",
    "    clear_output(wait=True)\n",
    "    try:\n",
    "        if top3_widget:\n",
    "            display(HTML('<p>query summary:</p>'),\n",
    "                    queries_df_widget,\n",
    "                    HTML('<p>top3:</p>'),\n",
    "                    top3_widget,\n",
    "                    HTML('<p>{} {}</p><p>{}</p>'.format(allele, name, query_string)),\n",
    "                    qgrid_widget)    \n",
    "        else:\n",
    "            display(queries_df_widget, HTML('<p>No results</p>'))\n",
    "    except Exception as e:          \n",
    "        print e        \n",
    "    \n",
    "queries_df_widget.observe(on_queries_df_row_selected, names=['_selected_rows'])   # ,  \n",
    "\n",
    "display(HTML('<p>query summary:</p>'),queries_df_widget)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def collect_md(s, md=None):\n",
    "    s = s.decode('UTF-8', 'ignore')\n",
    "    if md:\n",
    "        md = u'{}\\n{}'.format(md, s)\n",
    "    else:\n",
    "        md = u'{}'.format(s)\n",
    "    return md\n",
    "\n",
    "# report parameters\n",
    "query_names = ['alleles', '~location', '~range', 'protein_effects', 'protein_domain', '~biomarker_type',  'genes' , 'pathways']\n",
    "evidence_levels = ['A', 'B', 'C', 'D']\n",
    "# maximum number of matching evidence items\n",
    "limit = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "# p204\n",
       "\n",
       "## Drug Centric\n",
       "_____\n",
       "\n",
       "### ~location\n",
       "\n",
       "  * B mAb\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C SIROLIMUS\n",
       "    * cgi cancer ['SIROLIMUS']\n",
       "      * PTEN Sirolimus (MTOR inhibitor) Responsive\n",
       "      * [u'PTEN']\n",
       "      * [u'https://www.google.com/#q=ASCO 2013 (abstr 2532)']\n",
       "\n",
       "### ~range\n",
       "\n",
       "  * B mAb\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C EVEROLIMUS\n",
       "    * jax thyroid medullary carcinoma ['VANDETANIB', 'EVEROLIMUS']\n",
       "      * In a clinical case study, addition of Afinitor (everolimus)  to Caprelsa (vandetanib) treatment resulted in significant tumor reduction in a medullary thyroid carcinoma patient harboring ATM L804fs*4, ATM S978fs*12, and RET M918T, that achieved prolonged stable disease on Caprelsa (vandetanib) treatment alone (PMID: 27683183).\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27683183']\n",
       "    * jax thyroid carcinoma ['DABRAFENIB', 'EVEROLIMUS', 'TRAMETINIB']\n",
       "      * In a clinical case study, a patient with anaplastic thyroid carcinoma co-harboring BRAF V600E and PIK3CA H1047R demonstrated tumor regression when treated with the triple combination, Tafinlar (dabrafenib), Mekinist (trametinib), and Afinitor (everolimus) (PMID: 27797976).\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27797976']\n",
       "  * D TRAMETINIB\n",
       "    * jax melanoma ['TRAMETINIB']\n",
       "      * In a preclinical study, Mekinist (trametinib) reduced ERK signaling and inhibited proliferation of a melanoma cell line harboring BRAF G466E and HRAS Q61K in culture (PMID: 28783719).\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/28783719']\n",
       "    * jax melanoma ['TRAMETINIB']\n",
       "      * In a preclinical study, Mekinist (trametinib) in combination with Omipalisib (GSK2126458) resulted in improved growth inhibition in melanoma cell lines harboring BRAF V600E, NRAS Q61K, NRAS A146T and MAP2K1 P387S in culture, compared to either agent alone (PMID: 22389471).\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/22389471']\n",
       "\n",
       "### protein_effects\n",
       "\n",
       "  * B mAb\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C Sorafenib\n",
       "    * civic non-small cell lung carcinoma ['Sorafenib']\n",
       "      * In one patient with S214C mutation, the use of sorafenib has led to more than 5 years of survival and near remission.\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24569458']\n",
       "    * oncokb histiocytosis ['Sorafenib']\n",
       "      * Compelling clinical evidence supports the biomarker as being predictive of response to a drug in this indication but neither biomarker and drug are standard of care\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24569458', u'http://www.ncbi.nlm.nih.gov/pubmed/17016424', u'http://www.ncbi.nlm.nih.gov/pubmed/26566875']\n",
       "  * D Sorafenib\n",
       "    * civic non-small cell lung carcinoma ['Sorafenib']\n",
       "      * Sorafenib can inhibit ARF-mediated MEK phosphorylation and soft agar colony formation of AALE cells expressing ARAF S214C in vitro.\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24569458']\n",
       "\n",
       "### ~biomarker_type\n",
       "\n",
       "  * B GEFITINIB\n",
       "    * civic glioblastoma multiforme ['GEFITINIB', 'Erlotinib']\n",
       "      * Expression of EGFRvIII and PTEN was evaluated in 26 patients with glioblastomas treated with erlotinib or gefitinib and no EGFR or Her2/neu kinase domain mutations. 7 patients responded to treatment while 19 progressed rapidly. Coexpression of EGFRvIII and PTEN was associated with clinical response (defined as 25% tumor shrinkage). These results were confirmed using IHC in an independent cohort of 33 patients (8 responders) treated with erlotinib at another institution.\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/16282176']\n",
       "  * C EVEROLIMUS\n",
       "    * civic Her2-receptor positive breast cancer ['FULVESTRANT', 'EVEROLIMUS']\n",
       "      * Case report of a 44-year-old patient with recurrent HER2-positive breast cancer. The primary tumor was HR positive; however, the metastatic tumor was HR negative. The patient was resistant to classical chemotherapeutic agents and anti-HER2 treatment. Thus, the combination of everolimus and fulvestrant was administered. The patient experienced stable disease and PFS of 10 months. Genetic testing of pleural effusion after disease progression revealed a PIK3CA H1047R gene mutation but was positive for PTEN expression (IHC).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27445490']\n",
       "    * jax renal cell carcinoma ['EVEROLIMUS']\n",
       "      * In a retrospective analysis from a Phase II clinical trial, patients with metastatic renal cell carcinoma treated with Afinitor (everolimus) had no difference in progression free survival when stratified by high expression (n=17) and low expression (n=21) of Pten (PMID: 26951309).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26951309']\n",
       "  * D VEMURAFENIB\n",
       "    * jax melanoma ['VEMURAFENIB']\n",
       "      * In a preclinical study, expression of AKT1 E17K in a wild-type PTEN-expressing BRAF V600-mutant melanoma cell line conferred resistance to Zelboraf (vemurafenib) in culture (PMID: 24265152).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24265152']\n",
       "    * jax melanoma ['VEMURAFENIB']\n",
       "      * In a preclinical study, expression of AKT1 Q79K in a wild-type PTEN-expressing BRAF V600-mutant melanoma cell line conferred resistance to Zelboraf (vemurafenib) in culture (PMID: 24265152).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24265152']\n",
       "\n",
       "### genes\n",
       "\n",
       "  * B EVEROLIMUS\n",
       "    * civic endometrial carcinoma ['EVEROLIMUS']\n",
       "      * Tumors from 35 patients (12 with partial response or stable disease) with advanced endometrial carcinoma were assessed for PTEN mutational status, revealing 12 tumors with non-identical PTEN gene mutations. No mutations found in PTEN were effective in predicting tumor response to everolimus treatment. The clinical trial associated with this study is NCT00870337.\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/23238879']\n",
       "    * civic Her2-receptor positive breast cancer ['EVEROLIMUS']\n",
       "      * Archival tumor samples from 549 patients from the BOLERO-1 and -3 trials were analyzed for alterations in the PI3K pathway using next-generation sequencing, immunohistochemistry (IHC) and Sanger sequencing. In the BOLERO trials, patients with HER2-overexpressing advanced breast cancer were treated with trastuzumab, chemotherapy and randomized to receive either everolimus or placebo. In the pooled analysis of both trials, patients with PTEN loss (as defined by IHC staining intensity and proportion of positive tumor cells) had a longer progression-free survival with everolimus compared to placebo (HR 0.54 (95%CI 0.31 to 0.96, P=0.04) whereas patients with PTEN normal status had a hazard ration of 1.00 (95%CI 0.8 to 1.26, P=0.97).\n",
       "The same was found when PI3K mutations, PTEN loss and AKT E17K mutations were combined and analyzed as hyperactive PI3K pathway. Patients with PI3K hyperactive pathway had a hazard ratio of 0.67 (95%CI 0.48 to 0.93, P=0.02) compared to HR 1.19 (95%CI 0.87 to 1.62, P=0.28) in patients with PI3K normal pathway. These trends were independently confirmed in both trials but only pooled results showed consistent statistical significance.\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C EVEROLIMUS\n",
       "    * civic Her2-receptor positive breast cancer ['FULVESTRANT', 'EVEROLIMUS']\n",
       "      * Case report of a 44-year-old patient with recurrent HER2-positive breast cancer. The primary tumor was HR positive; however, the metastatic tumor was HR negative. The patient was resistant to classical chemotherapeutic agents and anti-HER2 treatment. Thus, the combination of everolimus and fulvestrant was administered. The patient experienced stable disease and PFS of 10 months. Genetic testing of pleural effusion after disease progression revealed a PIK3CA H1047R gene mutation but was positive for PTEN expression (IHC).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27445490']\n",
       "    * cgi prostate adenocarcinoma ['EVEROLIMUS']\n",
       "      * PTEN Everolimus (MTOR inhibitor) Responsive\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/23582881']\n",
       "  * D EVEROLIMUS\n",
       "    * civic breast cancer ['EVEROLIMUS', 'PP242']\n",
       "      * In a panel of 31 breast cancer cell lines, 11 had loss of PTEN function. Cells were tested for sensitivity to mTORC1 inhibitor everolimus and mTORC1/2 inhibitor PP242, and no significant correlation was found between PTEN status and sensitivity to mTORC inhibition.\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/21358673']\n",
       "    * jax breast cancer ['EVEROLIMUS', 'SELUMETINIB']\n",
       "      * In a preclinical study, breast cancer cells harboring PIK3CA wild-type and PTEN loss did not demonstrate any benefit when treated with a combination of Afinitor (everolimus) and Selumetinib (AZD6244) (PMID: 21358673).\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/21358673']\n",
       "\n",
       "### pathways\n",
       "\n",
       "  * B None\n",
       "  * C Imatinib\n",
       "    * jax sarcoma ['Imatinib']\n",
       "      * In a retrospective analysis, patients with desmoid fibromatosis harboring CTNNB1 S45F demonstrated a greater progression arrest rate at 6 months (85%) compared to patients with CTNNB1 wild-type (45%) when treated with Gleevec (imatinib) (PMID: 26861905). \n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26861905']\n",
       "    * jax sarcoma ['Imatinib']\n",
       "      * In a retrospective analysis, patients with desmoid fibromatosis harboring CTNNB1 T41A demonstrated a greater progression arrest rate at 6 months (70%) compared to patients with CTNNB1 wild-type (45%) when treated with Gleevec (imatinib) (PMID: 26861905). \n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26861905']\n",
       "  * D LGK\n",
       "    * molecularmatch rectal neoplasm ['LGK']\n",
       "      * CTNNB1 S37Y confers sensitivity to Lgk-974 in patients with Neoplasm of colorectum\n",
       "      * []\n",
       "      * [u'https://www.ncbi.nlm.nih.gov/pubmed/10347231', u'https://www.ncbi.nlm.nih.gov/pubmed/9065403']\n",
       "    * molecularmatch rectal neoplasm ['LGK']\n",
       "      * CTNNB1 T41A confers sensitivity to Lgk-974 in patients with Neoplasm of colorectum\n",
       "      * []\n",
       "      * [u'https://www.ncbi.nlm.nih.gov/pubmed/10698519', u'https://www.ncbi.nlm.nih.gov/pubmed/10487827', u'https://www.ncbi.nlm.nih.gov/pubmed/12200448']\n",
       "\n",
       "## Phenotype Centric\n",
       "_____\n",
       "\n",
       "### ~location\n",
       "\n",
       "  * B breast adenocarcinoma\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C prostate adenocarcinoma\n",
       "    * cgi prostate adenocarcinoma ['EVEROLIMUS']\n",
       "      * PTEN Everolimus (MTOR inhibitor) Responsive\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/23582881']\n",
       "\n",
       "### ~range\n",
       "\n",
       "  * B breast adenocarcinoma\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C melanoma\n",
       "    * jax melanoma ['VEMURAFENIB']\n",
       "      * In a clinical case study, a patient harboring BRAF V600E demonstrated an initial response to treatment with Zelboraf (vemurafenib), but progressed following emergence of a MAP2K1 C121S mutation (PMID: 21383288).\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/21383288']\n",
       "    * jax melanoma ['VEMURAFENIB']\n",
       "      * In a clinical study, a melanoma patient harboring a BRAF V600 mutation developed resistant mutations, MAP2K1 V60E, NRAS T58I, and NRAS Q61R, after 18 weeks of treatment with Zelboraf (vemurafenib) (PMID: 24265153).\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24265153']\n",
       "  * D colorectal cancer\n",
       "    * jax colorectal cancer ['PALBOCICLIB', 'TRAMETINIB']\n",
       "      * In a preclinical study, Mekinist (trametinib) and Ibrance (palbociclib)  combination treatment did not result in enhanced antitumor activity compared to Ibrance (palbociclib) alone in PDX models of KRAS wild-type colorectal cancer harboring BRAF V600E (PMID: 26369631).\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26369631']\n",
       "    * jax colorectal cancer ['VEMURAFENIB']\n",
       "      * In a preclinical study, colorectal cancer cells harboring BRAF V600E that acquired EGFR amplification and subsequent resistance to Selumetinib (AZD6244) and Zelboraf (vemurafenib) combination treatment were resistant to Zelboraf (vemurafenib) in culture (PMID: 27312529).\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27312529']\n",
       "\n",
       "### protein_effects\n",
       "\n",
       "  * B breast adenocarcinoma\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C prostate adenocarcinoma\n",
       "    * cgi prostate adenocarcinoma ['EVEROLIMUS']\n",
       "      * PTEN Everolimus (MTOR inhibitor) Responsive\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/23582881']\n",
       "    * cgi prostate adenocarcinoma ['EVEROLIMUS']\n",
       "      * PTEN Everolimus (MTOR inhibitor) Responsive\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/23582881']\n",
       "  * D non-small cell lung carcinoma\n",
       "    * civic non-small cell lung carcinoma ['Sorafenib']\n",
       "      * Sorafenib can inhibit ARF-mediated MEK phosphorylation and soft agar colony formation of AALE cells expressing ARAF S214C in vitro.\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24569458']\n",
       "    * civic non-small cell lung carcinoma ['TRAMETINIB']\n",
       "      * Trametinib can inhibit ARF-mediated ERK phosphorylation and soft agar colony formation of AALE cells expressing ARAF S214C in vitro.\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24569458']\n",
       "\n",
       "### ~biomarker_type\n",
       "\n",
       "  * B glioblastoma multiforme\n",
       "    * civic glioblastoma multiforme ['GEFITINIB', 'Erlotinib']\n",
       "      * Expression of EGFRvIII and PTEN was evaluated in 26 patients with glioblastomas treated with erlotinib or gefitinib and no EGFR or Her2/neu kinase domain mutations. 7 patients responded to treatment while 19 progressed rapidly. Coexpression of EGFRvIII and PTEN was associated with clinical response (defined as 25% tumor shrinkage). These results were confirmed using IHC in an independent cohort of 33 patients (8 responders) treated with erlotinib at another institution.\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/16282176']\n",
       "  * C renal cell carcinoma\n",
       "    * jax renal cell carcinoma ['EVEROLIMUS']\n",
       "      * In a retrospective analysis from a Phase II clinical trial, patients with metastatic renal cell carcinoma treated with Afinitor (everolimus) had no difference in progression free survival when stratified by high expression (n=17) and low expression (n=21) of Pten (PMID: 26951309).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26951309']\n",
       "    * jax renal cell carcinoma ['EVEROLIMUS']\n",
       "      * In a retrospective analysis from a Phase II clinical trial, patients with metastatic renal cell carcinoma treated with Afinitor (everolimus) had no difference in progression free survival when stratified by the presence (n=21) or absence (n=17) of PIK3CA activating mutations (PMID: 26951309).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26951309']\n",
       "  * D melanoma\n",
       "    * jax melanoma ['VEMURAFENIB']\n",
       "      * In a preclinical study, expression of AKT1 E17K in a wild-type PTEN-expressing BRAF V600-mutant melanoma cell line conferred resistance to Zelboraf (vemurafenib) in culture (PMID: 24265152).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24265152']\n",
       "    * jax melanoma ['VEMURAFENIB']\n",
       "      * In a preclinical study, expression of AKT1 Q79K in a wild-type PTEN-expressing BRAF V600-mutant melanoma cell line conferred resistance to Zelboraf (vemurafenib) in culture (PMID: 24265152).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24265152']\n",
       "\n",
       "### genes\n",
       "\n",
       "  * B prostate cancer\n",
       "    * civic prostate cancer ['EVEROLIMUS']\n",
       "      * In a phase II trial on castration-resistant prostate cancer patients, FISH detection of PTEN deletion was associated with increased progression free survival (p=0.07) and with improved PSA response (p=0.02) in response to everolimus, an mTOR pathway inhibitor.\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/23582881']\n",
       "  * C cancer\n",
       "    * civic Her2-receptor positive breast cancer ['FULVESTRANT', 'EVEROLIMUS']\n",
       "      * Case report of a 44-year-old patient with recurrent HER2-positive breast cancer. The primary tumor was HR positive; however, the metastatic tumor was HR negative. The patient was resistant to classical chemotherapeutic agents and anti-HER2 treatment. Thus, the combination of everolimus and fulvestrant was administered. The patient experienced stable disease and PFS of 10 months. Genetic testing of pleural effusion after disease progression revealed a PIK3CA H1047R gene mutation but was positive for PTEN expression (IHC).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27445490']\n",
       "    * jax cancer ['PEMETREXED', 'Sorafenib']\n",
       "      * In a Phase I clinical trial, low Pten expression or lack of Pten expression was associated with better response to treatment with the combination of Alimta (pemetrexed) and Nexavar (sorafenib) in patients with advanced solid tumors (PMID: 27213589).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27213589']\n",
       "  * D breast cancer\n",
       "    * civic breast cancer ['EVEROLIMUS', 'PP242']\n",
       "      * In a panel of 31 breast cancer cell lines, 11 had loss of PTEN function. Cells were tested for sensitivity to mTORC1 inhibitor everolimus and mTORC1/2 inhibitor PP242, and no significant correlation was found between PTEN status and sensitivity to mTORC inhibition.\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/21358673']\n",
       "    * jax breast cancer ['EVEROLIMUS', 'SELUMETINIB']\n",
       "      * In a preclinical study, breast cancer cells harboring PIK3CA wild-type and PTEN loss did not demonstrate any benefit when treated with a combination of Afinitor (everolimus) and Selumetinib (AZD6244) (PMID: 21358673).\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/21358673']\n",
       "\n",
       "### pathways\n",
       "\n",
       "  * B Desmoid Fibromatosis\n",
       "    * civic Desmoid Fibromatosis []\n",
       "      * In this study a collection of 254 sporadic desmoid tumors were analysed for presence of mutations in CTNNB1. A mutation in exon 3 of the CTNNB1 gene was identified in 223 (88%) of the 254 cases. The CTNNB1 mutation spectra concerned mainly three point mutations in two codons (41 and 45): T41A was identified in 132 cases (59%); S45F was identified in 49 samples (22%); and S45P was identified in 27 samples (12%). None of the 175 lesions mimicking desmoid tumors analyzed harbored any CTNNB1 mutations.\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/22766794']\n",
       "    * civic Desmoid Fibromatosis []\n",
       "      * Desmoid fibromatosis is a rare, nonmetastatic neoplasm marked by local invasiveness and relentless recurrence. Beta-Catenin deregulation has been commonly identified in sporadic desmoids. CTNNB1 mutations were observed in 117 of 138 (85%) of desmoids. Three discrete mutations in two codons of CTNNB1 exon 3 were identified: T41A (59%), S45F (33%), and S45P (8%). Five-year recurrence-free survival was significantly poorer in S45F-mutated desmoids (23%, P < 0.0001) versus either T41A (57%) or nonmutated tumors (65%). Because of their low frequency, patients with S45P mutations were excluded from these analyses but showed a recurrence course similar to T41A or wild type patients.\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/18832571']\n",
       "  * C sarcoma\n",
       "    * jax sarcoma ['Imatinib']\n",
       "      * In a retrospective analysis, patients with desmoid fibromatosis harboring CTNNB1 S45F demonstrated a greater progression arrest rate at 6 months (85%) compared to patients with CTNNB1 wild-type (45%) when treated with Gleevec (imatinib) (PMID: 26861905). \n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26861905']\n",
       "    * jax sarcoma ['Imatinib']\n",
       "      * In a retrospective analysis, patients with desmoid fibromatosis harboring CTNNB1 T41A demonstrated a greater progression arrest rate at 6 months (70%) compared to patients with CTNNB1 wild-type (45%) when treated with Gleevec (imatinib) (PMID: 26861905). \n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26861905']\n",
       "  * D rectal neoplasm\n",
       "    * molecularmatch rectal neoplasm ['LGK']\n",
       "      * CTNNB1 S37Y confers sensitivity to Lgk-974 in patients with Neoplasm of colorectum\n",
       "      * []\n",
       "      * [u'https://www.ncbi.nlm.nih.gov/pubmed/10347231', u'https://www.ncbi.nlm.nih.gov/pubmed/9065403']\n",
       "    * molecularmatch rectal neoplasm ['LGK']\n",
       "      * CTNNB1 T41A confers sensitivity to Lgk-974 in patients with Neoplasm of colorectum\n",
       "      * []\n",
       "      * [u'https://www.ncbi.nlm.nih.gov/pubmed/10698519', u'https://www.ncbi.nlm.nih.gov/pubmed/10487827', u'https://www.ncbi.nlm.nih.gov/pubmed/12200448']\n",
       "\n",
       "## Publication Centric\n",
       "_____\n",
       "\n",
       "### ~location\n",
       "\n",
       "  * B http://www.ncbi.nlm.nih.gov/pubmed/27091708\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C http://www.ncbi.nlm.nih.gov/pubmed/26861905\n",
       "    * jax sarcoma ['Imatinib']\n",
       "      * In a retrospective analysis, patients with desmoid fibromatosis harboring CTNNB1 T41A demonstrated a greater progression arrest rate at 6 months (70%) compared to patients with CTNNB1 wild-type (45%) when treated with Gleevec (imatinib) (PMID: 26861905). \n",
       "      * [u'CTNNB1', u'missense_variant', u'P35222', 'pathway']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26861905']\n",
       "\n",
       "### ~range\n",
       "\n",
       "  * B http://www.ncbi.nlm.nih.gov/pubmed/27091708\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C http://www.ncbi.nlm.nih.gov/pubmed/28514312\n",
       "    * jax colon cancer ['VEMURAFENIB', 'PANITUMUMAB']\n",
       "      * In a clinical study, a colon cancer patient harboring BRAF V600E and TP53 Q192K demonstrated a partial response when treated with a combination of Zelboraf (vemurafenib) and Vectibix (panitumumab) (PMID: 28514312). \n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/28514312']\n",
       "    * jax melanoma ['Imatinib', 'CARBOPLATIN', 'PACLITAXEL']\n",
       "      * In a clinical study, a melanoma patient harboring a PDGFRA mutation, CDKN2A mutation, MET exon 14 splice site mutation, and SMAD4 Q249H demonstrated a 21 month overall survival when treated with a combination of Gleevec (imatinib), Paraplatin (carboplatin), and Taxol (paclitaxel) (PMID: 28514312).\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/28514312']\n",
       "  * D http://www.ncbi.nlm.nih.gov/pubmed/27312529\n",
       "    * jax colorectal cancer ['VEMURAFENIB']\n",
       "      * In a preclinical study, colorectal cancer cells harboring BRAF V600E that acquired EGFR amplification and subsequent resistance to Selumetinib (AZD6244) and Zelboraf (vemurafenib) combination treatment were resistant to Zelboraf (vemurafenib) in culture (PMID: 27312529).\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27312529']\n",
       "    * jax colorectal cancer ['VEMURAFENIB', 'CETUXIMAB']\n",
       "      * In a preclinical study, colorectal cancer cells harboring BRAF V600E that acquired KRAS amplification and subsequent resistance to Erbitux (cetuximab) and Tafinlar (dabrafenib) combination treatment were resistant to combination therapy consisting of Erbitux (cetuximab) and Zelboraf (vemurafenib) in culture (PMID: 27312529).\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27312529']\n",
       "\n",
       "### protein_effects\n",
       "\n",
       "  * B http://www.ncbi.nlm.nih.gov/pubmed/27091708\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C http://www.ncbi.nlm.nih.gov/pubmed/24569458\n",
       "    * civic non-small cell lung carcinoma ['Sorafenib']\n",
       "      * In one patient with S214C mutation, the use of sorafenib has led to more than 5 years of survival and near remission.\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24569458']\n",
       "    * oncokb histiocytosis ['Sorafenib']\n",
       "      * Compelling clinical evidence supports the biomarker as being predictive of response to a drug in this indication but neither biomarker and drug are standard of care\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24569458', u'http://www.ncbi.nlm.nih.gov/pubmed/17016424', u'http://www.ncbi.nlm.nih.gov/pubmed/26566875']\n",
       "  * D http://www.ncbi.nlm.nih.gov/pubmed/24569458\n",
       "    * civic non-small cell lung carcinoma ['Sorafenib']\n",
       "      * Sorafenib can inhibit ARF-mediated MEK phosphorylation and soft agar colony formation of AALE cells expressing ARAF S214C in vitro.\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24569458']\n",
       "    * civic non-small cell lung carcinoma ['TRAMETINIB']\n",
       "      * Trametinib can inhibit ARF-mediated ERK phosphorylation and soft agar colony formation of AALE cells expressing ARAF S214C in vitro.\n",
       "      * [u'missense_variant']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24569458']\n",
       "\n",
       "### ~biomarker_type\n",
       "\n",
       "  * B http://www.ncbi.nlm.nih.gov/pubmed/16282176\n",
       "    * civic glioblastoma multiforme ['GEFITINIB', 'Erlotinib']\n",
       "      * Expression of EGFRvIII and PTEN was evaluated in 26 patients with glioblastomas treated with erlotinib or gefitinib and no EGFR or Her2/neu kinase domain mutations. 7 patients responded to treatment while 19 progressed rapidly. Coexpression of EGFRvIII and PTEN was associated with clinical response (defined as 25% tumor shrinkage). These results were confirmed using IHC in an independent cohort of 33 patients (8 responders) treated with erlotinib at another institution.\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/16282176']\n",
       "  * C http://www.ncbi.nlm.nih.gov/pubmed/26951309\n",
       "    * jax renal cell carcinoma ['EVEROLIMUS']\n",
       "      * In a retrospective analysis from a Phase II clinical trial, patients with metastatic renal cell carcinoma treated with Afinitor (everolimus) had no difference in progression free survival when stratified by high expression (n=17) and low expression (n=21) of Pten (PMID: 26951309).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26951309']\n",
       "    * jax renal cell carcinoma ['EVEROLIMUS']\n",
       "      * In a retrospective analysis from a Phase II clinical trial, patients with metastatic renal cell carcinoma treated with Afinitor (everolimus) had no difference in progression free survival when stratified by the presence (n=21) or absence (n=17) of PIK3CA activating mutations (PMID: 26951309).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26951309']\n",
       "  * D http://www.ncbi.nlm.nih.gov/pubmed/24265152\n",
       "    * jax melanoma ['VEMURAFENIB']\n",
       "      * In a preclinical study, expression of AKT1 E17K in a wild-type PTEN-expressing BRAF V600-mutant melanoma cell line conferred resistance to Zelboraf (vemurafenib) in culture (PMID: 24265152).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24265152']\n",
       "    * jax melanoma ['VEMURAFENIB']\n",
       "      * In a preclinical study, expression of AKT1 Q79K in a wild-type PTEN-expressing BRAF V600-mutant melanoma cell line conferred resistance to Zelboraf (vemurafenib) in culture (PMID: 24265152).\n",
       "      * [u'PTEN', u'Uncategorized']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/24265152']\n",
       "\n",
       "### genes\n",
       "\n",
       "  * B http://www.ncbi.nlm.nih.gov/pubmed/27091708\n",
       "    * civic Her2-receptor positive breast cancer ['EVEROLIMUS']\n",
       "      * Archival tumor samples from 549 patients from the BOLERO-1 and -3 trials were analyzed for alterations in the PI3K pathway using next-generation sequencing, immunohistochemistry (IHC) and Sanger sequencing. In the BOLERO trials, patients with HER2-overexpressing advanced breast cancer were treated with trastuzumab, chemotherapy and randomized to receive either everolimus or placebo. In the pooled analysis of both trials, patients with PTEN loss (as defined by IHC staining intensity and proportion of positive tumor cells) had a longer progression-free survival with everolimus compared to placebo (HR 0.54 (95%CI 0.31 to 0.96, P=0.04) whereas patients with PTEN normal status had a hazard ration of 1.00 (95%CI 0.8 to 1.26, P=0.97).\n",
       "The same was found when PI3K mutations, PTEN loss and AKT E17K mutations were combined and analyzed as hyperactive PI3K pathway. Patients with PI3K hyperactive pathway had a hazard ratio of 0.67 (95%CI 0.48 to 0.93, P=0.02) compared to HR 1.19 (95%CI 0.87 to 1.62, P=0.28) in patients with PI3K normal pathway. These trends were independently confirmed in both trials but only pooled results showed consistent statistical significance.\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "    * cgi breast adenocarcinoma ['EVEROLIMUS', 'TRASTUZUMAB', 'mAb']\n",
       "      * PTEN ERBB2 Everolimus + Trastuzumab + Chemotherapy (MTOR inhibitor + ERBB2 mAb inhibitor + Chemotherapy) Responsive\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/27091708']\n",
       "  * C http://www.ncbi.nlm.nih.gov/pubmed/26861905\n",
       "    * jax sarcoma ['Imatinib']\n",
       "      * In a retrospective analysis, patients with desmoid fibromatosis harboring CTNNB1 S45F demonstrated a greater progression arrest rate at 6 months (85%) compared to patients with CTNNB1 wild-type (45%) when treated with Gleevec (imatinib) (PMID: 26861905). \n",
       "      * [u'CTNNB1', u'missense_variant', u'P35222', 'pathway']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26861905']\n",
       "    * jax sarcoma ['Imatinib']\n",
       "      * In a retrospective analysis, patients with desmoid fibromatosis harboring CTNNB1 activating mutations demonstrated a greater progression arrest rate at 6 months compared to patients with CTNNB1 wild-type when treated with Gleevec (imatinib) (PMID: 26861905).\n",
       "      * [u'CTNNB1']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26861905']\n",
       "  * D http://www.ncbi.nlm.nih.gov/pubmed/21358673\n",
       "    * civic breast cancer ['EVEROLIMUS', 'PP242']\n",
       "      * In a panel of 31 breast cancer cell lines, 11 had loss of PTEN function. Cells were tested for sensitivity to mTORC1 inhibitor everolimus and mTORC1/2 inhibitor PP242, and no significant correlation was found between PTEN status and sensitivity to mTORC inhibition.\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/21358673']\n",
       "    * jax breast cancer ['EVEROLIMUS', 'SELUMETINIB']\n",
       "      * In a preclinical study, breast cancer cells harboring PIK3CA wild-type and PTEN loss did not demonstrate any benefit when treated with a combination of Afinitor (everolimus) and Selumetinib (AZD6244) (PMID: 21358673).\n",
       "      * [u'PTEN']\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/21358673']\n",
       "\n",
       "### pathways\n",
       "\n",
       "  * B http://www.ncbi.nlm.nih.gov/pubmed/18832571\n",
       "    * civic Desmoid Fibromatosis []\n",
       "      * Desmoid fibromatosis is a rare, nonmetastatic neoplasm marked by local invasiveness and relentless recurrence. Beta-Catenin deregulation has been commonly identified in sporadic desmoids. CTNNB1 mutations were observed in 117 of 138 (85%) of desmoids. Three discrete mutations in two codons of CTNNB1 exon 3 were identified: T41A (59%), S45F (33%), and S45P (8%). Five-year recurrence-free survival was significantly poorer in S45F-mutated desmoids (23%, P < 0.0001) versus either T41A (57%) or nonmutated tumors (65%). Because of their low frequency, patients with S45P mutations were excluded from these analyses but showed a recurrence course similar to T41A or wild type patients.\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/18832571']\n",
       "    * civic Desmoid Fibromatosis []\n",
       "      * Desmoid fibromatosis is a rare, nonmetastatic neoplasm marked by local invasiveness and relentless recurrence. Beta-Catenin deregulation has been commonly identified in sporadic desmoids.\n",
       "CTNNB1 mutations were observed in 117 of 138 (85%) of desmoids. Three discrete mutations in two codons of CTNNB1 exon 3 were identified: T41A (59%), S45F (33%), and S45P (8%). In conclusion, CTNNB1 mutations are highly common in desmoid tumors.\n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/18832571']\n",
       "  * C http://www.ncbi.nlm.nih.gov/pubmed/26861905\n",
       "    * jax sarcoma ['Imatinib']\n",
       "      * In a retrospective analysis, patients with desmoid fibromatosis harboring CTNNB1 S45F demonstrated a greater progression arrest rate at 6 months (85%) compared to patients with CTNNB1 wild-type (45%) when treated with Gleevec (imatinib) (PMID: 26861905). \n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26861905']\n",
       "    * jax sarcoma ['Imatinib']\n",
       "      * In a retrospective analysis, patients with desmoid fibromatosis harboring CTNNB1 T41A demonstrated a greater progression arrest rate at 6 months (70%) compared to patients with CTNNB1 wild-type (45%) when treated with Gleevec (imatinib) (PMID: 26861905). \n",
       "      * []\n",
       "      * [u'http://www.ncbi.nlm.nih.gov/pubmed/26861905']\n",
       "  * D https://www.ncbi.nlm.nih.gov/pubmed/10347231\n",
       "    * molecularmatch rectal neoplasm ['LGK']\n",
       "      * CTNNB1 S37Y confers sensitivity to Lgk-974 in patients with Neoplasm of colorectum\n",
       "      * []\n",
       "      * [u'https://www.ncbi.nlm.nih.gov/pubmed/10347231', u'https://www.ncbi.nlm.nih.gov/pubmed/9065403']\n",
       "    * molecularmatch rectal neoplasm ['LGK']\n",
       "      * CTNNB1 S37F confers sensitivity to Lgk-974 in patients with Neoplasm of colorectum\n",
       "      * []\n",
       "      * [u'https://www.ncbi.nlm.nih.gov/pubmed/10347231', u'https://www.ncbi.nlm.nih.gov/pubmed/11196159', u'https://www.ncbi.nlm.nih.gov/pubmed/11943721']"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    " \n",
    "md = collect_md('# {}'.format(dataset_name))\n",
    "\n",
    "md = collect_md('\\n## Drug Centric\\n_____', md)\n",
    "\n",
    "for name in query_names:\n",
    "    df = pd.DataFrame()\n",
    "    for allele in data_frames.keys():\n",
    "        if name in data_frames[allele] and len(data_frames[allele][name]):\n",
    "            df = df.append(data_frames[allele][name])\n",
    "    if len(df) == 0:\n",
    "        continue\n",
    "    md = collect_md('\\n### {}\\n'.format(name), md)\n",
    "    for evidence_level in evidence_levels:\n",
    "        evidence_level_df = df[df.evidence_label == evidence_level]\n",
    "        if len(evidence_level_df):\n",
    "            drug = top(evidence_level_df, fields=['drugs'], size=1)[0]['value']\n",
    "            md = collect_md('  * {} {}'.format(evidence_level, drug), md)\n",
    "            c = 0\n",
    "            for evidence in evidence_level_df.to_records():\n",
    "                if drug in evidence.drugs:\n",
    "                    md = collect_md('    * {} {} {}'.format(evidence.source, evidence.phenotype, evidence.drugs), md)\n",
    "                    md = collect_md('      * {}'.format(evidence.description), md)\n",
    "                    md = collect_md('      * {}'.format(evidence.matches), md)                    \n",
    "                    md = collect_md('      * {}'.format(evidence.publications), md)                                        \n",
    "                    c += 1\n",
    "                    if c == limit:\n",
    "                        break\n",
    "                        \n",
    "md = collect_md('\\n## Phenotype Centric\\n_____', md)\n",
    "\n",
    "for name in query_names:\n",
    "    df = pd.DataFrame()\n",
    "    for allele in data_frames.keys():\n",
    "        if name in data_frames[allele] and len(data_frames[allele][name]):\n",
    "            df = df.append(data_frames[allele][name])\n",
    "    if len(df) == 0:\n",
    "        continue\n",
    "    md = collect_md('\\n### {}\\n'.format(name), md)\n",
    "    for evidence_level in evidence_levels:\n",
    "        evidence_level_df = df[df.evidence_label == evidence_level]\n",
    "        if len(evidence_level_df):\n",
    "            phenotype = top(evidence_level_df, fields=['phenotype'], size=1)[0]['value']\n",
    "            md = collect_md('  * {} {}'.format(evidence_level, phenotype), md)\n",
    "            c = 0\n",
    "            for evidence in evidence_level_df.to_records():\n",
    "                if phenotype in evidence.phenotype:\n",
    "                    md = collect_md('    * {} {} {}'.format(evidence.source, evidence.phenotype, evidence.drugs), md)\n",
    "                    md = collect_md('      * {}'.format(evidence.description), md)\n",
    "                    md = collect_md('      * {}'.format(evidence.matches), md)                    \n",
    "                    md = collect_md('      * {}'.format(evidence.publications), md)                                        \n",
    "                    c += 1\n",
    "                    if c == limit:\n",
    "                        break\n",
    "\n",
    "                        \n",
    "                        \n",
    "md = collect_md('\\n## Publication Centric\\n_____', md)\n",
    "\n",
    "for name in query_names:\n",
    "    df = pd.DataFrame()\n",
    "    for allele in data_frames.keys():\n",
    "        if name in data_frames[allele] and len(data_frames[allele][name]):\n",
    "            df = df.append(data_frames[allele][name])\n",
    "    if len(df) == 0:\n",
    "        continue\n",
    "    md = collect_md('\\n### {}\\n'.format(name), md)\n",
    "    for evidence_level in evidence_levels:\n",
    "        evidence_level_df = df[df.evidence_label == evidence_level]\n",
    "        if len(evidence_level_df):\n",
    "            publication = top(evidence_level_df, fields=['publications'], size=1)[0]['value']\n",
    "            md = collect_md('  * {} {}'.format(evidence_level, publication), md)\n",
    "            c = 0\n",
    "            for evidence in evidence_level_df.to_records():\n",
    "                if publication in evidence.publications:\n",
    "                    md = collect_md('    * {} {} {}'.format(evidence.source, evidence.phenotype, evidence.drugs), md)\n",
    "                    md = collect_md('      * {}'.format(evidence.description), md)\n",
    "                    md = collect_md('      * {}'.format(evidence.matches), md)                   \n",
    "                    md = collect_md('      * {}'.format(evidence.publications), md)                    \n",
    "                    c += 1\n",
    "                    if c == limit:\n",
    "                        break\n",
    "\n",
    "                        \n",
    "                        \n",
    "# print md    \n",
    "Markdown(md)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
